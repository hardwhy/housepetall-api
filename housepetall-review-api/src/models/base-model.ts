export class BaseModel {
  createdAt?: string;
  createdBy?: string;
}