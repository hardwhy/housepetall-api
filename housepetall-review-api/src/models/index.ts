export * from './base-model';
export * from './review';