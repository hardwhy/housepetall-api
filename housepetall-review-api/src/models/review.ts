import { BaseModel } from "./base-model";

export interface Review extends BaseModel{
  id: number;
  name: string;
  petName: string;
  rating: number;
  comments?: string;
}
