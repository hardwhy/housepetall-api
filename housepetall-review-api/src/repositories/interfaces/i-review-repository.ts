import { Review } from '../../models';

export interface IReviewRepository {
  create(review: Review): Promise<Review>;
  findAll(): Promise<Review[]>;
}
