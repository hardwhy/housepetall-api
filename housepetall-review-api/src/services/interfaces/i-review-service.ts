import { Review } from '../../models/review';

export interface IReviewService {
  submitReview(data: Review): Promise<Review>;
  getAllReviews(): Promise<Review[]>;
}
