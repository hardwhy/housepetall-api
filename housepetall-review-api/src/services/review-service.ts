import { Review } from '../models/review';
import { IReviewService } from './interfaces/i-review-service';
import { IReviewRepository } from '../repositories/interfaces/i-review-repository';

export class ReviewService implements IReviewService {
  constructor(private repository: IReviewRepository) {}

  async submitReview(data: Review): Promise<Review> {
    const createdAt = new Date().toISOString().split('T')[0];
    const review: Review = {
      ...data,
      createdAt: createdAt,
      updatedAt: createdAt,
    };

    return await this.repository.create(review);
  }
}
