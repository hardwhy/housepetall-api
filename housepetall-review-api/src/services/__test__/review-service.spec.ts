import { Review } from '../../models/review';
import { IReviewRepository } from '../../repositories/interfaces/i-review-repository';
import { ReviewService } from '../review-service';

describe('ReviewService Test', () => {
  const reviewRepository: IReviewRepository = {
    create: jest.fn(),
    findAll: jest.fn(),
  };

  const service = new ReviewService(reviewRepository);

  const createdAt = new Date().toISOString().split('T')[0];
  const review: Review = {
    id: 'id',
    name: 'name',
    petName: 'petName',
    rating: 5,
    comments: 'comments',
  };

  const reviewWithCreatedAt: Review = {
    ...review,
    createdAt: createdAt,
    updatedAt: createdAt,
  };

  describe('submitReview', () => {
    it('should return submitted review', async () => {
      //Given
      reviewRepository.create = jest
        .fn()
        .mockResolvedValue(reviewWithCreatedAt);
      //When
      const result = await service.submitReview(review);

      expect(result).toEqual(reviewWithCreatedAt);
      expect(result.createdAt).toEqual(createdAt);
      expect(result.updatedAt).toEqual(createdAt);
      expect(reviewRepository.create).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });
  });
  describe('getAllReviews', () => {
    it('should return all created review', async () => {
      //Given
      reviewRepository.findAll = jest
        .fn()
        .mockResolvedValue([reviewWithCreatedAt]);

      //When
      const result = await service.getAllReviews();

      expect(result).toEqual([reviewWithCreatedAt]);
      expect(result.length).toEqual(1);
      expect(reviewRepository.findAll).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });
  });
});
