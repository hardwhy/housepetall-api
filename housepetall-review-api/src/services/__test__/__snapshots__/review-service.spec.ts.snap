// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewService Test getAllReviews should return all created review 1`] = `
[
  {
    "comments": "comments",
    "createdAt": "2025-02-20",
    "id": "id",
    "name": "name",
    "petName": "petName",
    "rating": 5,
    "updatedAt": "2025-02-20",
  },
]
`;

exports[`ReviewService Test submitReview should return submitted review 1`] = `
{
  "comments": "comments",
  "createdAt": "2025-02-20",
  "id": "id",
  "name": "name",
  "petName": "petName",
  "rating": 5,
  "updatedAt": "2025-02-20",
}
`;
